gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java

```diff
@@ -363,20 +363,47 @@ public Void call() throws Exception {
     return workUnitFile.toString();
   }
 
-  private void waitForJobCompletion() throws InterruptedException {
-    while (true) {
+  private void waitForJobCompletion()  throws InterruptedException {
+    LOGGER.info("Waiting for job to complete...");
+    long timeoutInSeconds;
+    timeoutInSeconds = Long.parseLong(this.jobProps.getProperty(ConfigurationKeys.HELIX_JOB_TIMEOUT_SECONDS,
+        ConfigurationKeys.DEFAULT_HELIX_JOB_TIMEOUT_SECONDS));
+    long endTime = System.currentTimeMillis() + timeoutInSeconds*1000;
+    while (System.currentTimeMillis() <= endTime) {
       WorkflowContext workflowContext = TaskDriver.getWorkflowContext(this.helixManager, this.helixQueueName);
       if (workflowContext != null) {
         org.apache.helix.task.TaskState helixJobState = workflowContext.getJobState(this.jobResourceName);
-        if (helixJobState == org.apache.helix.task.TaskState.COMPLETED ||
-            helixJobState == org.apache.helix.task.TaskState.FAILED ||
-            helixJobState == org.apache.helix.task.TaskState.STOPPED) {
+        if (helixJobState == org.apache.helix.task.TaskState.COMPLETED
+            || helixJobState == org.apache.helix.task.TaskState.FAILED
+            || helixJobState == org.apache.helix.task.TaskState.STOPPED) {
           return;
         }
       }
-
       Thread.sleep(1000);
     }
+    helixTaskDriverWaitToStop(this.helixQueueName, 10L);
+    startCancellationExecutor();
```

Comment:
Shouldn't this be started before the stop call? Also, why don't we always start this for this job launcher? cancelJob() may be called in other instances.